### PR TITLE
Add a module API to send an HTTP push notification

### PR DESCRIPTION
**Pull request by MatMaul**
This is useful so you can wake up a device and send it some data through push, in the case where for some reason this data is not a Matrix event per itself.

Possible use cases could be to be sure that a device is logged out immediately when deleting a device, or a phone home feature for enterprise needs.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/module_api/__init__.py [hunk 1]
```diff
@@ -1226,6 +1228,38 @@ async def sleep(self, seconds: float) -> None:
...
+
+        Args:
+            user_id: The user ID of the device where to send the push notification.
+            device_id: The device ID of the device where to send the push notification.
+            content: A dict of values that will be put in the `notification` field of the push
+            (cf Push Gatway spec). `devices` field will be overrided if included.
+            tweaks: A dict of `tweaks` that will be inserted in the `devices` section, cf spec.
+
+        Returns:
+            True if at least one push was succesfully sent, False in case of error or if no
```
Comment (MatMaul) [inline synapse/module_api/__init__.py:1251]: I think we probably want to distinguish the 2 types of error here (no pusher vs fail) ?
Or do we want to just ignore if there is no pusher ?

I am thinking about returning false for the no pusher case, and raising an exception for the failure.

File: synapse/module_api/__init__.py [hunk 2]
```diff
@@ -1226,6 +1229,43 @@ async def sleep(self, seconds: float) -> None:
...
+            user_id: The user ID to send the push notification to.
+            device_id: The device ID of the device where to send the push notification. If `None`,
+            the notification will be sent to all registered HTTP pushers of the user.
+            content: A dict of values that will be put in the `notification` field of the push
+            (cf Push Gatway spec). `devices` field will be overrided if included.
+            tweaks: A dict of `tweaks` that will be inserted in the `devices` section, cf spec.
+
+        Returns:
+            True if at least one push was succesfully sent, False if no matching pusher has been
+            found or an error occured when sending the push.
```
Comment (MatMaul) [inline synapse/module_api/__init__.py:1254]: Perhaps we should differentiate between no pusher found and errors. Also currently if we don't specify `device_id` and some pusher fails but at least one succeeded, this is still returning success.

File: synapse/module_api/__init__.py [hunk 3]
```diff
@@ -1226,6 +1228,43 @@ async def sleep(self, seconds: float) -> None:
...
+        for the specified user/device.
+
+        Added in Synapse v1.82.0.
+
+        Args:
+            user_id: The user ID to send the push notification to.
+            device_id: The device ID of the device where to send the push notification. If `None`,
+            the notification will be sent to all registered HTTP pushers of the user.
+            content: A dict of values that will be put in the `notification` field of the push
+            (cf Push Gatway spec). `devices` field will be overrided if included.
```
Comment (clokep) [inline synapse/module_api/__init__.py:1248]:
```suggestion
            (cf Push Gateway spec). `devices` field will be overrided if included.
```

File: synapse/push/httppusher.py [hunk 4]
```diff
@@ -342,9 +344,12 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
 
-    async def _build_notification_dict(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
-    ) -> Dict[str, Any]:
+    async def _build_event_notification(
+        self,
+        event: EventBase,
+        tweaks: JsonMapping,
+        badge: int,
+    ) -> Tuple[JsonDict, JsonMapping]:
```
Comment (clokep) [inline synapse/push/httppusher.py:352]: Please add a docstring.

File: synapse/push/httppusher.py [hunk 5]
```diff
@@ -358,80 +363,70 @@ async def _build_notification_dict(
...
+            "app_id": self.app_id,
+            "pushkey": self.pushkey,
+            "pushkey_ts": int(self.pushkey_ts / 1000),
+            "data": self.data_minus_url,
+        }
+        if tweaks:
+            device["tweaks"] = tweaks
 
-        return d
+        content["devices"] = [device]
```
Comment (clokep) [inline synapse/push/httppusher.py:417]: This modifies an input parameter. This might be OK, but should be called out somewhere.

File: synapse/push/httppusher.py [hunk 6]
```diff
@@ -358,80 +363,70 @@ async def _build_notification_dict(
...
+        self, content: JsonDict, tweaks: Optional[JsonMapping]
+    ) -> JsonDict:
+        device = {
+            "app_id": self.app_id,
+            "pushkey": self.pushkey,
+            "pushkey_ts": int(self.pushkey_ts / 1000),
+            "data": self.data_minus_url,
+        }
+        if tweaks:
+            device["tweaks"] = tweaks
```
Comment (clokep) [inline synapse/push/httppusher.py:415]: This was previously not set for `event_id_only`, I think? This is a change in behavior.

File: synapse/push/httppusher.py [hunk 7]
```diff
@@ -358,80 +363,70 @@ async def _build_notification_dict(
...
 
-        return d
+        content["devices"] = [device]
+
+        return {"notification": content}
 
     async def dispatch_push(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
+        self, content: JsonDict, tweaks: Optional[JsonMapping] = None
     ) -> Union[bool, Iterable[str]]:
```
Comment (clokep) [inline synapse/push/httppusher.py:423]: While we're here, please add a docstring.

File: synapse/push/httppusher.py [hunk 8]
```diff
@@ -440,10 +435,27 @@ async def dispatch_push(
...
 
+    async def dispatch_push_event(
+        self,
+        event: EventBase,
+        tweaks: JsonMapping,
+        badge: int,
+    ) -> Union[bool, Iterable[str]]:
+        content, tweaks = await self._build_event_notification(event, tweaks, badge)
+        if not content:
+            return []
```
Comment (clokep) [inline synapse/push/httppusher.py:448]: I don't think `content` can ever be false-y?

File: synapse/module_api/__init__.py (hunk 2 above)
Comment (clokep) [reply synapse/module_api/__init__.py:1253]: Maybe return a map of `device_id -> success`?

File: synapse/module_api/__init__.py [hunk 9]
```diff
@@ -1226,6 +1228,43 @@ async def sleep(self, seconds: float) -> None:
 
         await self._clock.sleep(seconds)
 
+    async def send_http_push_notification(
+        self,
+        user_id: str,
+        device_id: Optional[str],
+        content: JsonDict,
+        tweaks: Optional[JsonMapping] = None,
+    ) -> bool:
```
Comment (clokep) [inline synapse/module_api/__init__.py:1237]: This seems OK, but I question whether it is reasonable to expose this. It is insanely low-level for a module API.

**Review by clokep — COMMENTED** [review]
I'm finding this PR quite hard to follow -- it would be helpful if the refactoring was done piecemeal in separate commits (or as separate PRs).

Sorry to ask you to add lots of docstrings, but if we're here it would be quite helpful to improve the code!

File: synapse/push/httppusher.py (hunk 6 above)
Comment (MatMaul) [reply synapse/push/httppusher.py:415]: It will be clearer after adding some docstrings, but `_build_event_notification` takes `tweaks` and return `tweaks`, which will be empty for the `event_id_only` use case.

File: synapse/push/httppusher.py [hunk 10]
```diff
@@ -342,9 +344,28 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
+        Args:
+            event: the event
+            tweaks: tweaks to apply
+            badge: unread count to send with the push notification
+
+        Returns: a tuple `(content, tweaks)`, where:
+           * `content` is the content to put in `notification` before sending it to the
+           push gateway.
+           * `tweaks` is the tweaks to put in the `devices` section, they could be
+           different from the ones inputed in this function.
```
Comment (clokep) [inline synapse/push/httppusher.py:367]:
```suggestion
        Returns:
            A tuple of:
                The content to put in `notification` before sending it to the push gateway.

                The tweaks to put in the `devices` section, they may be different from the input.
```

File: synapse/push/httppusher.py [hunk 11]
```diff
@@ -342,9 +344,28 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
+    async def _build_event_notification_content(
+        self,
+        event: EventBase,
+        tweaks: JsonMapping,
+        badge: int,
+    ) -> Tuple[JsonDict, JsonMapping]:
+        """Build the content of a push notification from an event, as specified by the
+        Push Gateway spec. This excludes the top level `notification` element and
+        is not taking care of the `devices` section either, those are handled by
+        `_build_notification_dict`.
```
Comment (clokep) [inline synapse/push/httppusher.py:356]:
```suggestion
        """
        Build the content of a push notification from an event, as specified by the
        Push Gateway spec.

        _build_notification_dict uses this to build the notification field and handles
        adding the devices field.
```

File: synapse/push/httppusher.py [hunk 12]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+        """Build a full notification from the content of it, as specified by the
+        Push Gateway spec. It wraps the content in a top level `notification`
+        element and put relevant device info in the `devices` section.
+
+        Args:
+            content: the content
+            tweaks: tweaks to add into the `devices` section
+            default_payload: default payload to add in `devices[0].data.default_payload`.
+            This will be merged (and override if some matching values already exist there)
+            with existing `default_payload`.
```
Comment (clokep) [inline synapse/push/httppusher.py:436]:
```suggestion
            default_payload: default payload to add in `devices[0].data.default_payload`.
                This will be merged (and override if some matching values already exist there)
                with existing `default_payload`.
```

File: synapse/push/httppusher.py [hunk 13]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+        Returns:
+            a full notification that can be send to the registered push gateway.
+        """
+        content = content.copy()
+
+        data = self.data_minus_url.copy()
+        if default_payload:
+            if "default_payload" not in data:
+                data["default_payload"] = {}
+            data["default_payload"].update(default_payload)
```
Comment (clokep) [inline synapse/push/httppusher.py:447]:
```suggestion
            data.setdefault("default_payload", {}).update(default_payload)
```

File: synapse/push/httppusher.py [hunk 14]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+
+        content["devices"] = [device]
+
+        return {"notification": content}
 
     async def dispatch_push(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
+        self,
+        content: JsonDict,
+        tweaks: Optional[JsonMapping] = None,
```
Comment (clokep) [inline synapse/push/httppusher.py:465]: This significantly expands the type of `tweaks` -- is that on purpose? It seems incorrect.

File: synapse/push/httppusher.py [hunk 15]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+        default_payload: Optional[JsonMapping] = None,
     ) -> Union[bool, Iterable[str]]:
-        notification_dict = await self._build_notification_dict(event, tweaks, badge)
-        if not notification_dict:
-            return []
+        """Send a notification to the registered push gateway, with `content` being
+        the content of the `notification` top property specified in the spec.
+        If specified `devices` property will be overrided with device-specific
+        information for this pusher.
```
Comment (clokep) [inline synapse/push/httppusher.py:471]:
```suggestion
        If specified, the `devices` property will be overridden with device-specific
        information for this pusher.
```

File: synapse/push/httppusher.py [hunk 16]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+
+        Args:
+            content: the content
+            tweaks: tweaks to add into the `devices` section
+            default_payload: default payload to add in `devices[0].data.default_payload`.
+            This will be merged (and override if some matching values already exist there)
+            with existing `default_payload`.
```
Comment (clokep) [inline synapse/push/httppusher.py:478]:
```suggestion
            default_payload: default payload to add in `devices[0].data.default_payload`.
                This will be merged (and override if some matching values already exist there)
                with existing `default_payload`.
```

File: synapse/push/httppusher.py [hunk 17]
```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
...
+
+        Returns:
+            False if an error occured when calling the push gateway, or an array of
+            rejected push keys otherwise. If this array is empty, the push fully
+            succeeded.
+        """
+        notif_dict = self._build_notification_dict(content, tweaks, default_payload)
```
Comment (clokep) [inline synapse/push/httppusher.py:485]: Would it be clearer if we inlined `_build_notification_dict` here?

File: synapse/push/httppusher.py [hunk 18]
```diff
@@ -440,10 +496,41 @@ async def dispatch_push(
...
+            badge: unread count to send with the push notification
+
+        Returns:
+            False if an error occured when calling the push gateway, or an array of
+            rejected push keys otherwise. If this array is empty, the push fully
+            succeeded.
+        """
+        content, tweaks = await self._build_event_notification_content(
+            event, tweaks, badge
+        )
```
Comment (clokep) [inline synapse/push/httppusher.py:523]: I suspect this code would be simpler if we inlined `_build_event_notification_content` here? Or is it separate for testing or something?

I'm finding it quite difficult to follow the logic; I think because the tweaks go into a separate piece (under `devices`), but whether we include them or not depends on the internsls of `_build_event_notification_content`?

File: synapse/push/httppusher.py (hunk 14 above)
Comment (MatMaul) [reply synapse/push/httppusher.py:465]: I think the original type hint is wrong. The values can at least have `bool` and `str` from what I understand from the code and the example in the spec. The spec itself doesn't define `tweaks` :/ .
`JsonMapping` is probably too broad, I just checked and I should probably use `Mapping[str, SimpleJsonValue]` I think.

File: synapse/push/httppusher.py (hunk 17 above)
Comment (MatMaul) [reply synapse/push/httppusher.py:485]: you are right, I am going to inline them. I kept it because it was there and didn't question it, but the `dispatch` methods are simple enough so I think it makes sense to inline.

File: synapse/module_api/__init__.py (hunk 9 above)
Comment (MatMaul) [reply synapse/module_api/__init__.py]: I've put a rational and some examples in the top comment.

File: synapse/push/httppusher.py [hunk 19]
```diff
@@ -56,7 +57,9 @@
 )
 
 
-def tweaks_for_actions(actions: List[Union[str, Dict]]) -> Dict[str, Any]:
+def tweaks_for_actions(
+    actions: List[Union[str, Dict]]
+) -> Mapping[str, SimpleJsonValue]:
```
Comment (clokep) [inline synapse/push/httppusher.py:62]: I don't see anythign stopping this from being any type of value? Why using `SimpleJsonValue` OK here?

Comment (clokep) [reply synapse/push/httppusher.py:62]: Let's use the other thread.

File: synapse/push/httppusher.py (hunk 14 above)
Comment (clokep) [reply synapse/push/httppusher.py]: I'm pretty sure we don't do validation of what tweaks you set so this could be anything, but I'm not 100% sure. Where do we build this dictionary that makes it safe to make this change?

Comment (MatMaul) [reply synapse/push/httppusher.py]: So original code was using `Dict[str, bool]`, and I am broadening the type, because clearly the type hint is wrong: [tweak_for_actions](https://github.com/matrix-org/synapse/blob/develop/synapse/push/httppusher.py#L59) is returning a `Dict[str, Any]`.
I am not really understanding your point sorry :) I was using `JsonMapping` but moved to `Mapping[str, SimpleJsonValue]` after your concern about being too broad, but I think we should use `JsonMapping` and I am going to change that again since it seems you agree :) ?

Comment (clokep) [reply synapse/push/httppusher.py]: Alright, I see what you're saying. I agree that `JsonMapping` is likely the correct thing to use. To be clear -- I didn't know where this value came from, so pointing to `tweak_for_actions` is helpful in understanding what generates this dictionary.

File: synapse/push/httppusher.py (hunk 19 above)
Comment (clokep) [reply synapse/push/httppusher.py:62]: OK, so this needs to be broader since we're not limiting it. I think either leaving this as is or using `JsonMapping` is probably OK.

File: synapse/push/httppusher.py [hunk 20]
```diff
@@ -342,9 +346,83 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
                     await self._pusherpool.remove_pusher(self.app_id, pk, self.user_id)
         return True
 
-    async def _build_notification_dict(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
-    ) -> Dict[str, Any]:
+    async def dispatch_push(
+        self,
+        content: JsonDict,
+        tweaks: Optional[Mapping[str, SimpleJsonValue]] = None,
```
Comment (clokep) [inline synapse/push/httppusher.py:352]: I think this is meant to be `Optional[JsonMapping]`?

File: synapse/push/httppusher.py [hunk 21]
```diff
@@ -342,9 +346,83 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
+        default_payload: Optional[JsonMapping] = None,
+    ) -> Union[bool, Iterable[str]]:
+        """Send a notification to the registered push gateway, with `content` being
+        the content of the `notification` top property specified in the spec.
+        If specified, the `devices` property will be overridden with device-specific
+        information for this pusher.
```
Comment (clokep) [inline synapse/push/httppusher.py:358]: I'm tripping over the "If specified" which makes it sound like it is sometimes valid to provide this and sometimes not

```suggestion
        Note that the `devices` property will be added with device-specific
        information for this pusher.
```

File: synapse/push/httppusher.py [hunk 22]
```diff
@@ -342,9 +346,83 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
+        tweaks: Mapping[str, SimpleJsonValue],
+        badge: int,
+    ) -> Union[bool, Iterable[str]]:
+        """Send a notification to the registered push gateway by building it
+        from an event.
+
+        Args:
+            event: the event
+            tweaks: tweaks to add into the `devices` section, and to decide the
+            priority to use
```
Comment (clokep) [inline synapse/push/httppusher.py:418]:
```suggestion
            tweaks: tweaks to add into the `devices` section, used to decide the
                priority to use
```

File: synapse/push/httppusher.py [hunk 23]
```diff
@@ -392,57 +459,28 @@ async def _build_notification_dict(
...
             return False
-        rejected = []
-        if "rejected" in resp:
-            rejected = resp["rejected"]
-        if not rejected:
+        if not res:
             self.badge_count_last_call = badge
-        return rejected
+
+        return res
```
Comment (clokep) [inline synapse/push/httppusher.py:483]: This is a bit hard to follow, maybe:

```python
# If the push is successful and none are rejected, update the badge count.
if res is not False and not res:
    self.bad_count_last_call = badge
return res
```

File: synapse/push/httppusher.py [hunk 24]
```diff
@@ -101,6 +104,7 @@ def __init__(self, hs: "HomeServer", pusher_config: PusherConfig):
         self._storage_controllers = self.hs.get_storage_controllers()
         self.app_display_name = pusher_config.app_display_name
         self.device_display_name = pusher_config.device_display_name
+        self.device_id = pusher_config.device_id
```
Comment (clokep) [inline synapse/push/httppusher.py:107]: I think this is unused?

File: synapse/push/httppusher.py [hunk 25]
```diff
@@ -342,9 +346,83 @@ async def _process_one(self, push_action: HttpPushAction) -> bool:
...
 
-    async def _build_notification_dict(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
-    ) -> Dict[str, Any]:
+    async def dispatch_push(
+        self,
+        content: JsonDict,
+        tweaks: Optional[Mapping[str, SimpleJsonValue]] = None,
+        default_payload: Optional[JsonMapping] = None,
+    ) -> Union[bool, Iterable[str]]:
```
Comment (clokep) [inline synapse/push/httppusher.py:354]:
```suggestion
    ) -> Union[bool, List[str]]:
```

File: synapse/module_api/__init__.py [hunk 26]
```diff
@@ -1226,6 +1228,52 @@ async def sleep(self, seconds: float) -> None:
...
+        if user_id in self._pusherpool.pushers:
+            for p in self._pusherpool.pushers[user_id].values():
+                if isinstance(p, HttpPusher) and (
+                    not device_id or p.device_id == device_id
+                ):
+                    sent = False
+                    res = await p.dispatch_push(content, tweaks, default_payload)
+                    if (
+                        isinstance(res, (list, tuple)) and len(res) == 0
+                    ) or res is not False:
```
Comment (clokep) [inline synapse/module_api/__init__.py:1268]:
```suggestion
                    # If the push was successful and no pushers were rejected.
                    if res is not False and not res:
```

File: synapse/module_api/__init__.py [hunk 27]
```diff
@@ -1226,6 +1228,52 @@ async def sleep(self, seconds: float) -> None:
...
+                        sent = True
```
Comment (clokep) [inline synapse/module_api/__init__.py:1269]:
```suggestion
                    res = await p.dispatch_push(content, tweaks, default_payload)
                    # Check if the push was successful and no pushers were rejected.
                    sent = res is not False and not res
```

File: synapse/push/httppusher.py (hunk 24 above)
Comment (MatMaul) [reply synapse/push/httppusher.py:105]: It is but in the `ModuleApi`, not inside the `HttpPusher`.

File: synapse/module_api/__init__.py [hunk 28]
```diff
@@ -1225,6 +1227,51 @@ async def sleep(self, seconds: float) -> None:
...
+        Returns:
+            a dict reprensenting the status of the push per device ID
+        """
+        status = {}
+        if user_id in self._pusherpool.pushers:
+            for p in self._pusherpool.pushers[user_id].values():
+                if isinstance(p, HttpPusher) and (
+                    not device_id or p.device_id == device_id
+                ):
+                    sent = False
```
Comment (clokep) [inline synapse/module_api/__init__.py:1263]: This is now unused:

```suggestion
```

**Review by clokep — APPROVED** [review]
